source/iopipe/json/formatter.d

```diff
@@ -0,0 +1,641 @@
+
+import std.bitmanip : BitArray;
+import std.exception : enforce;
+import std.traits : isNumeric;
+import std.conv : to;
+import std.regex;
+import std.algorithm.iteration : substitute;
+import std.algorithm.searching : startsWith, endsWith;
+import std.format;
+
+struct MemberOptions {
+    char quoteChar = '"';        // Use " or ' for quotes
+    bool escapeKey = true;       // Whether to escape special chars in key
+    
+    enum ColonSpacing {
+        none,  // "key":"value"
+        after, // "key": "value"
+        both,  // "key" : "value"
+    }
+    
+    ColonSpacing colonSpacing = ColonSpacing.after;  // Default spacing
+}
+
+enum KeywordValue : string {
+    Null = "null",
+    True = "true",
+    False = "false",
+    // JSON5 extensions
+    Infinity = "Infinity",
+    NegativeInfinity = "-Infinity",
+    NaN = "NaN"
+}
+
+struct JSONFormatter(Chain) 
+if (isIopipe!Chain)
+{
+private:
+
+    Chain* outputChain;
+    int spacing = 0;
+    int indent = 4;
+    MemberOptions memberOptions;
+
+    size_t nWritten = 0;
+    char currentQuoteChar = char.init;
+
+    private enum State : ubyte
+    {
+        Begin,  // next item should be either an Object or Array or String
+        First,  // Expect the first member of a new object or array.
+        /*
+        * In an object:
+        *   - First
+        *     - Value(key)
+        *   - Value(value)
+        *   
+        *   - Member
+        *     - Value(key)
+        *   - Value(value)
+        *   ...
+        *
+        * In an array:
+        *   - First(value)
+        *   - Member(value)
+        *   ...
+        */
+        Member, // Expect next member (key for object, value for array)
+        Value,  // Expect value
+        End     // there shouldn't be any more items
+    }
+
+    // 0 = array, 1 = object (same convention as parser)
+    BitArray stack;
+    size_t   stackLen;
+    State    state = State.Begin;
+
+    bool inObj()  const @property nothrow
+    {
+        return stackLen == 0 ? false : stack[stackLen - 1];
+    }
+
+    // Can we add a value (string, number, keyword, begin object/array)?
+    bool canAddValue() const nothrow
+    {
+        final switch (state) with (State)
+        {
+        case Begin:
+            return true;  // root value
+        case First:
+            // only allowed for arrays; for objects we must go through addMember
+            return !inObj();
+        case Member:
+            // only allowed for arrays; for objects we must go through addMember
+            return !inObj();
+        case Value:
+            return true;
+        case End:
+            return false;
+        }
+    }
+
+    // Can we add an object member (key)?
+    // If so, add comma/indent as needed.
+    bool canAddObjMember()
+    {
+        if (!inObj())
+            return false;
+        final switch (state) with (State)
+        {
+        case First:
+            // no need to change state to Member, any key string added later will do that
```

Comment:
Since we are no longer changing states ever in these, this comment is now out of place.